bundler/lib/bundler.rb

```diff
@@ -39,6 +39,16 @@ module Bundler
   environment_preserver.replace_with_backup
   SUDO_MUTEX = Thread::Mutex.new
 
+  SAFE_MARSHAL_CLASSES = [Symbol, TrueClass, String, Array, Hash].freeze
```

Comment:
:thinking: I'm not sure why `TrueClass` is needed, but it is passed to proc even not part of the structure.

```ruby
irb(main):013:0> simple_structure = { "name" => [:abc] }
=> {"name"=>[:abc]}
irb(main):014:0> Marshal.load(Marshal.dump(simple_structure), ->(object) { object.tap { puts object.class }})
TrueClass
String
Symbol
Array
Hash
=> {"name"=>[:abc]}
```

Reply:
I was able to strip this down to string deserialization.

```ruby
Marshal.load(Marshal.dump('string'), ->(o) { o.tap { puts o.class } })
String
TrueClass
```

Reply:
In case it's useful, I think `TrueClass` is related to the encoding. That seems to be what the ET at the end of the dump is about:

```rb
Marshal.dump('string')
#=> "\x04\bI\"\vstring\x06:\x06ET"
```

whereas we get `false` in this case:

```rb
Marshal.dump('string.encode!('ASCII'))
#=> "\x04\bI\"\vstring\x06:\x06EF"
```

or a string for other encodings:

```rb
Marshal.dump("string".encode!('ISO-8859-1'))
#=> "\x04\bI\"\vstring\x06:\rencoding\"\x0FISO-8859-1"
```